Add copy button to out-of-band code page

TimMnz09 opened the pull request:
## Description

This PR adds a copy button to the out-of-band code page to improve user experience when copying authentication codes.

## Changes

- Added a copy button with clipboard functionality to the code.ftl template
- Implemented modern clipboard API with fallback to execCommand for broader browser compatibility
- Added visual feedback showing Copied confirmation when the button is clicked
- Button is temporarily disabled after copying to prevent multiple rapid clicks

## Motivation

Currently, the out-of-band page displays a success code and instructs users to copy and paste it into their application, but there is no copy button provided. This can be inconvenient for users and may lead to errors when manually selecting and copying the code.

## Testing

The implementation includes:
- Modern navigator.clipboard API for secure contexts
- Fallback to document.execCommand for older browsers
- Visual feedback with icon change and temporary button disable
- Proper accessibility attributes

<img width="591" height="357" alt="image" src="https://github.com/user-attachments/assets/84336a9b-3f00-4858-96ec-6f80577d27cc" />

Closes #42094

ahus1 commented:
Thank you for this screenshot. Looking at the very long code that a user can't actually see fully, an input which is only one line is not a good fit. Maybe a well-formatted span or div with some text in it that wraps would be a better fit. WDYT? Again, this might be a follow-up thing as it wasn't part of the original issue.  

Looping in @jonkoops to have an opinion here on the JavaScript, and optionally on the idea of not having an input field here.

TimMnz09 committed:
Add copy functionality to out-of-band code page

- Add clipboard copy functionality for oob code
- Use styling from official patternfly component
- Improved user experience with visual feedback on copy action

Closes #42094

Signed-off-by: Tim Menze <58325404+TimMnz09@users.noreply.github.com>

jonkoops commented:
Thanks for the PR @TimMnz09, I totally agree we should provide a better experience here by offering a clipboard functionality.

I would prefer to use the built-in PatternFly components here as much as possible, and avoid adding custom styling. This makes the code more portable, and easier to customize when theming, as well as making things easier for us when we eventually need to upgrade to the next version of PatternFly.

Thankfully, PatternFly offers a [clipboard copy](https://v5-archive.patternfly.org/components/clipboard-copy) component that we can use out of the box, so let's use that here. We should strive to match the behavior of the React component here, with a tooltip when the copy succeeds. I think we can try to place this component in `field.ftl`, but worst case, it could be contained in its own template.

We expect that Keycloak is run in a secure context in any production scenario, so simply feature detecting the Clipboard API should be enough, removing the copy button if not present. Additionally, we do not support the legacy browser that might use `execCommand()` to copy text to the clipboard, so this fallback can be safely removed.

jonkoops commented:
Thanks! I will take a look at your proposed changes.

jonkoops commented on themes/src/main/resources/theme/keycloak.v2/login/code.ftl:
```diff
@@ -11,10 +11,82 @@
+                    
+                    function showCopySuccess(button, originalIcon) {
+                        // Show success feedback with tooltip-like behavior
+                        button.innerHTML = '<i class="fas fa-check" aria-hidden="true"></i>';
```
If you need to use some HTML, preferably it should go in a `<template>` tag, co-located with the original template. This allows for easier customization by custom theme authors, without then also needing to update this template with the JavaScript.

TimMnz09 replied:
I completely reworked this. Instead of using template of inline html, I also switched to `data-` labels.

jonkoops commented on themes/src/main/resources/theme/keycloak.v2/login/code.ftl:
```diff
@@ -11,10 +11,82 @@
+                        button.disabled = true;
+                        
+                        // Reset button after 2 seconds
+                        setTimeout(function() {
```
```suggestion
                        setTimeout(() => {
```

jonkoops commented on themes/src/main/resources/theme/keycloak.v2/login/field.ftl:
```diff
@@ -88,6 +88,54 @@
 
+<#macro clipboard name label value="" readonly=true>
+  <@group name=name label=label>
+    <div class="pf-v5-c-clipboard-copy" id="${name}-clipboard">
```
Like other classes, e.g. `kcInputGroupItemClass`, this class should be provided by a property to allow easy customization. This applies for all `pf-` classes.

jonkoops commented on themes/src/main/resources/theme/keycloak.v2/login/field.ftl:
```diff
@@ -88,6 +88,54 @@
+            <button 
+              class="${properties.kcFormPasswordVisibilityButtonClass}" 
+              type="button" 
+              aria-label="Show content"
```
This label should be translated.

jonkoops reviewed (approved):
Looks great, thanks!

ahus1 committed:
Review

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 reviewed (approved):
As one of the tests was failing, I've reverted the ID of the input field to the original ID of the field. This will also help others who are running automation tasks against Keycloak if it remains constant. 

Also did a tiny bit of CSS to wrap at any character as this seems to be the most appropriate way to break a code.

jonkoops commented on themes/src/main/resources/theme/keycloak.v2/login/resources/css/styles.css:
```diff
@@ -109,6 +109,10 @@ div.kc-logo-text span {
     text-align: center;
 }
 
+#kc-code pre code {
+    word-break: break-all;
+}
```
Rather than introducing custom CSS, I would prefer we use [PatternFly utility classes](https://v5-archive.patternfly.org/utility-classes/text#controlling-text) where possible.
